Created ReviewRemindersDatabase and ReviewReminder

ericli3690 opened the pull request:
## Purpose / Description
Review Reminders database system.

- Created `ReviewRemindersDatabase`, which contains methods for reading and writing to a Preferences Datastore instance for storing review reminder data
- Created `ReviewReminder`, which defines the review reminder data class
- Created Robolectric unit tests for `ReviewRemindersDatabase` and `ReviewReminder`

## Fixes
- For GSoC 2025: Review Reminders

## Approach
Using Preferences Datastore was the original method I proposed for storing review reminder data in my original project proposal: see [[1]](https://docs.google.com/document/d/1KMjnm6ol1M_VMAsg9V5Qb4XRZnEON-39gkzLkFe8q2Y/edit?tab=t.0#heading=h.o472euuadswv)

Why did I initially choose Preferences Datastore for storing review reminder data?
1. Google recommends using it over Shared Preferences: see [[2]](https://android-developers.googleblog.com/2020/09/prefer-storing-data-with-jetpack.html)
2. It has been used before. Past attempts at creating a review reminders system (see [[3]](https://github.com/ankidroid/Anki-Android/pull/11487)) used Preferences Datastore.

However, we've now chosen to use SharedPreferences instead. It's consistent with what the rest of AnkiDroid uses and doesn't necessitate a whole new import. It's also a lot simpler and straightforward.

## How Has This Been Tested?
- `ReviewRemindersDatabaseTest` directly or indirectly exercises all methods within `ReviewRemindersDatabase`.
- `ReviewReminderTest` exercises the ID allocation method of `ReviewReminder`
- On a physical Samsung S23, API 34., buttons to add, edit, and delete reminders (coming soon in a later PR) cause expected changes to the database and the database changes persist between app sessions.

## Learning
I've tried my best to make my code clean and safe. In particular, allocating and deallocating IDs is handled internally and any developer creating a new `ReviewReminder` should be unable to manually forge an ID.

Originally, I planned to create different types of review reminders. I've now abandoned that in favour of a simpler approach involving snooze functionality.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

ericli3690 commented:
Hi, here's a first draft of what I'd like the review reminder storage local database storage system to be like. Looking forward to constructive criticism. For now, I’ve put these files in anki/notifications since that seemed a reasonable place to put them. Let me know if I should move them elsewhere, such as by creating a new anki/reviewreminders folder.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/notifications/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,224 @@
+    ) {
+        val reminders = getRemindersForDeck(did).toMutableList()
+        val indexToEdit = reminders.indexOfFirst { it.id == reminderId }
+        if (indexToEdit == -1) {
```
So long as the rest of my code that I write over the next few months is not faulty, the failure clauses of `editReviewReminderForDeck` and `deleteReviewReminderForDeck` should never execute. It just felt a bit wrong to let the two methods die with a NPE rather than throwing a more expressive error, so I included these two error clauses. I don't think we need specific UI/UX to handle these cases as they should never occur. Should I get rid of these error throws?

ericli3690 commented:
Had to do a second force push because I forgot to move the files to a new package as outlined at #18318.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,77 @@
+ */
+@Serializable
+@ConsistentCopyVisibility
+data class ReviewReminder private constructor(
```
I think it should be called "ReviewReminderSettings" or something like that. Because it does not represents a reminedr. It represents the setting that is used for a reminder by day.

This name seems to indicate that this corresponds to a very specific reminder. That is, one specific message shown at a specific point in time to the user

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Retrieve the list of reminders for a specific deck.
```
You don't need to document it's a list, this is clear from the type. Instead, just, note that it returns the reminder settings for this specific deck.

It's important to note you return the settings and not the reminders. Because maybe at the time the code is executed there is 0 actual reminder that exists. But the settings still exists

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+     * @return The list of reminders associated with this key.
+     */
+    private suspend fun getRemindersForKey(key: Preferences.Key<String>): List<ReviewReminder> {
+        val jsonString =
+            context.dataStore.data.firstOrNull()?.let {
+                it[key]
+            } ?: ""
+        if (jsonString.isEmpty()) return emptyList()
```
I don't understand why, if it's null, you use an empty string instead of directly returning

```kotlin
        val jsonString =
            context.dataStore.data.firstOrNull()?.let {
                it[key]
            } ?: return emptyList()
``` 
seems better. Unless there is a reason why you expect that the value associated to this preference exists but is the empty string. If so, you should document how it could occur

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+    }
+
+    /**
+     * Allocate and return the next free reminder ID which can be associated with a new review reminder.
```
Please be more specific. I have on idea of what "allocate" mean.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+        // Get next free ID and previously used IDs
+        val nextFreeIdKey = intPreferencesKey(DATASTORE_NEXT_FREE_ID_KEY)
+        val previouslyUsedIdsKey = stringPreferencesKey(DATASTORE_PREVIOUSLY_USED_IDS_KEY)
+        val (nextFreeId, previouslyUsedIds) =
```
I'd prefer 
```kotlin
        val (nextFreeId, previouslyUsedIdsList) =
            context.dataStore.data.firstOrNull()?.let { preferences ->
                val nextFreeId = preferences[nextFreeIdKey] ?: FIRST_REMINDER_ID
                val previouslyUsedIds = preferences[previouslyUsedIdsKey]?.let { Json.decodeFromString<MutableList<Int>>(it) } ?: mutableListOf()
                Pair(nextFreeId, previouslyUsedIds)
            } ?: Pair(FIRST_REMINDER_ID, mutableListOf())
```

ericli3690 replied:
This has been removed. See below comment for more information.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+            } ?: Pair(FIRST_REMINDER_ID, "[]")
+        val previouslyUsedIdsList = Json.decodeFromString<MutableList<Int>>(previouslyUsedIds)
+
+        // Use previously used IDs if available, else increment next free ID
```
Why? I really don't understand the reason for this.

There are 2^32 ints. Maybe 2^31-1 if you want to use non negative numbers. We can just have a counter, save the value of the latest assigned id and just increment it each time.
Not only is this unrealistic that someone create billions of reminder and delete them, because of the time it would take. But even if it was instantaneous, if you save all deleted ints, it means you'd have a list of billion of ints in your datastore. That could not work

ericli3690 replied:
Ah, you're right. I initially went with this complicated method because it felt "wrong" to just increment a counter with each new added notification and just let that grow as far as it wanted.

Removing this overly-complicated allocation functionality makes ReviewRemindersDatabase a lot shorter, too, which is a nice plus. Implemented!

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+        reminderId: Int,
+    ) {
+        val reminders = getRemindersForDeck(did).toMutableList()
+        val reminderToRemove = reminders.find { it.id == reminderId }
+        if (reminderToRemove == null) {
+            throw IllegalArgumentException("Delete Reminder: Reminder with ID $reminderId not found in deck $did")
+        }
+        reminders.remove(reminderToRemove)
```
I believe you can simplify that to 
```kotlin
        val removed = reminders.removeIf { it.id == reminderId }
        if (!removed) {
            throw IllegalArgumentException("Delete Reminder: Reminder with ID $reminderId not found in deck $did")
        }
```

ericli3690 replied:
I've fully removed the addReviewReminder, editReviewReminder, and deleteReviewReminder methods and instead plan to move them to the onClick methods of the UI buttons for adding, editing, and deleting ReviewReminders respectively, since that's probably the only place adding, editing, and deleting code will ever be called anyways. Putting them into ReviewRemindersDatabase just makes ReviewRemindersDatabase more cluttered and messy in my opinion.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+     * @param reminderId The ID of the reminder to delete.
+     */
+    suspend fun deleteReviewReminderForDeck(
+        did: Long,
```
Please, use DeckId as a type. Here and everywhere

ericli3690 replied:
Oops, I'll make sure to do this every time going forward.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+     * @param did The deck ID.
+     * @return The list of reminders.
+     */
+    suspend fun getRemindersForDeck(did: Long): List<ReviewReminder> {
```
Is there a reason for it not to be private? I assume it's for test. If so, here and elsewhere, use VisibleForTest

ericli3690 replied:
The `get` and `set` functions are intentionally public, yes. I plan on using them from the ScheduleReminders UI. I might mark them as internal at a later date if I find that I don't need to refer to them from any other place.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+        reminderId: Int,
+        updatedReminder: ReviewReminder,
+    ) {
+        val reminders = getRemindersForDeck(did).toMutableList()
```
I feel like you have a lot of code repetition here. 
As you always use `toMutableList` maybe the function should directly return  a mutable list (and note that mutating the input returned won't be registered unless it's through `setRemindersForDeck`)

You can also have an helper function that ensures as follow

```kotlin
    private suspend fun editReminder(did: DeckId, reminderEditor: (MutableList<ReviewReminder> )->List<ReviewReminder>) {
        setRemindersForDeck(did, reminderEditor(getRemindersForDeck(did)))
    }

    /**
     * Add a review reminder for a specific deck.
     * @param did The deck ID.
     * @param newReminder The new review reminder to add.
     */
    suspend fun addReviewReminderForDeck(
        did: Long,
        newReminder: ReviewReminder,
    ) {
        editReminder(did) {reminders ->
            reminders.add(newReminder)
            return@editReminder reminders
        }
    }
```

And reuse `editReminder` in both method below

ericli3690 replied:
Ah, I like this a lot, thanks. Implemented!

Arthur-Milchior commented on AnkiDroid/src/test/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabaseTest.kt:
```diff
@@ -0,0 +1,195 @@
+/*
+ *  Copyright (c) 2025 Eric Li <ericli3690@gmail.com>
+ *
```
I believe it's usually better to have test in the same commit as what is tested. At least if it's already written.

ericli3690 replied:
Sounds good. I've merged all my commits together.

ericli3690 commented:
Repeated force push because I forgot to rebase, my bad.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,77 @@
+
+/**
+ * Review reminder data class. Handles the fields of a review reminders and the logic behind creating one.
+ * Below, a public way of creating [ReviewReminder] is exposed via a companion object so that IDs are not abused.
```
I've now clarified it a bit more. I was referring to developers creating overlapping IDs, creating IDs out of order, etc. In general we want IDs to increment in a nice orderly fashion without developer interference.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,77 @@
+    companion object {
+        /**
+         * Create a new review reminder. This will allocate a new ID for the reminder.
+         * @param context The context to use for Datastore database access.
```
Sounds good. I've now removed @params annotations wherever I feel like they are self-explanatory.

ericli3690 commented on AnkiDroid/build.gradle:
```diff
@@ -370,6 +370,7 @@ dependencies {
     implementation libs.androidx.appcompat
     implementation libs.androidx.browser
     implementation libs.androidx.core.ktx
+    implementation libs.androidx.datastore.preferences
```
Datastore has now been removed in favour of SharedPrefs. 👍

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+import kotlinx.serialization.json.Json
+
+/**
+ * Preferences Datastore for storing review reminders.
```
Datastore has now been removed in favour of SharedPrefs. 👍

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+/**
+ * Preferences Datastore for storing review reminders.
+ */
+internal val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "review_reminders")
```
Datastore has now been removed in favour of SharedPrefs. 👍
I've also moved things around so nothing is marked as internal anymore, such as by moving ID allocation to `ReviewReminder`. Everything is either public or private.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,234 @@
+         * - next_free_id: Int: The next available free ID for a reminder
+         * - previously_used_ids: String (JSON of List<Int>): JSON string of the list of previously used but now available IDs
+         */
+        private const val DATASTORE_DECK_SPECIFIC_KEY = "deck_%s"
```
Writing an entire new private method just for string formatting when `String.format` already exists feels like a bit much, though. I personally like the concise syntax this little `%s`allows me to use, for instance:

```getRemindersForKey(String.format(DATASTORE_DECK_SPECIFIC_KEY, did))```

And doing the following makes the declaration of DATASTORE_DECK_SPECIFIC_KEY a little harder to understand at a glance in my opinion:

```
private const val DATASTORE_DECK_SPECIFIC_KEY = "deck_"
// ...
getRemindersForKey(DATASTORE_DECK_SPECIFIC_KEY + did)
```

Was there a different method you had in mind?

ericli3690 replied:
I've changed my mind, this method:
```
private const val DECK_SPECIFIC_KEY = "review_reminder_deck_"
// ...
getRemindersForKey(DECK_SPECIFIC_KEY + did)
```
...is better, as it also allows me to filter SharedPrefs to find all deck-specific reminders for use in `getAllDeckSpecificReminders`. Implemented.

ericli3690 commented:
Hi, I've refactored to use SharedPreferences instead of Preferences Datastore. In retrospect, thanks for telling me to use it, it really is a lot more simple and I should have started with this method. Using Datastore was like trying to crack a nut with a sledgehammer. The only negative consequence is that my branch I'm trying to merge these changes in from is still called `review-reminders-datastore`, so that's a bit confusing-looking but shouldn't be an issue.

Regarding other overly-complex points I've simplified: the fancy ID-allocation system is now gone, and I've decided to move the addReminderForDeck, editReminderForDeck, and deleteReminderForDeck methods to the UI onClick methods since that should be the only place they're called from, and so they don't need to be in ReviewRemindersDatabase.

I've also done my best to clean up the docstrings, as per feedback. Previously they were more ad-hoc, now I've tried my best to make them genuine pieces of documentation.

ericli3690 commented:
I realized the SharedPreferences keys are too short. Since there's tons of other things being stored in SharedPreferences, it's probably a good idea to append `review_reminders_` to the beginning of the keys I use in these files so they don't collide with any other keys.

ericli3690 commented:
I realized while I was building out the ScheduleReminders RecyclerView more that:
1. There are typos in this commit, whoops.
2. I need to store a review reminder's corresponding deck ID within ReviewReminder for the RecyclerView to work properly in GLOBAL_DECK_SPECIFIC scope mode.
3. I need to add a method to retrieve the reminders in all decks to ReviewRemindersDatastore, to be used in GLOBAL_DECK_SPECIFIC scope mode.

ericli3690 commented:
Rebased and fixed a List that should have been a MutableList.

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/reviewreminders/ReviewReminderTest.kt:
```diff
@@ -0,0 +1,57 @@
+    fun `getAndIncrementNextFreeReminderId should increment IDs correctly`() {
+        for (i in 0..10) {
+            val reminder = ReviewReminder.createReviewReminder(activity, ReviewReminderTypes.SINGLE, 12, 30, 5)
+            assert(reminder.id == i)
```
Use Hamcrest (`assertThat`) as it provides a better error message

Don't use `assert` in tests. In particular don't use a boolean, as that returns pass/fail, rather than why a failure occurred.

ericli3690 commented:
- Added a ReviewReminderId typealias.
- Tweaked the contracts of the database's "edit" methods to now require a lambda which returns a List rather than strictly a MutableList.
- Added an editAllDeckSpecificReminders method. I realized I needed it for the deck-specific scope mode's editing while creating the bulk deletion feature.
- Added a new test to ReviewRemindersDatabaseTest to test editAllDeckSpecificReminders.
- Cleaned up some syntax in the two new test files.

ericli3690 commented:
Please do not review this PR. I'm actively refactoring it as I've realized I can streamline the UX. Sorry for asking for reviews and then going back and changing things again, I really do appreciate the constructive criticism of my code and am sorry for pulling it out of the merge pipeline when it's so close to the end.

ericli3690 commented:
Refactored. Removed ReviewReminderTypes and replaced it with snoozing. Looking forward to reviews!

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,137 @@
+        val allReminders = mutableListOf<ReviewReminder>()
+        context.sharedPrefs().all.forEach { (key, value) ->
+            if (key.startsWith(DECK_SPECIFIC_KEY)) {
+                val reminders = Json.decodeFromString<MutableList<ReviewReminder>>(value.toString())
```
What happens if the JSON string is corrupted?

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * Edit all [ReviewReminder]s that are associated with a specific deck by operating on a single mutable list.
+     */
+    fun editAllDeckSpecificReminders(reminderEditor: (MutableList<ReviewReminder>) -> List<ReviewReminder>) {
+        val newReminders = reminderEditor(getAllDeckSpecificReminders())
+        val remindersForEachDeck: Map<DeckId, List<ReviewReminder>> = newReminders.groupBy { it.did }
+
+        // Clear existing review reminder keys in SharedPreferences
+        val existingKeys =
+            context
+                .sharedPrefs()
+                .all.keys
+                .filter { it.startsWith(DECK_SPECIFIC_KEY) }
+
+        context.sharedPrefs().edit {
+            existingKeys.forEach { remove(it) }
+            // Add the updated ones back in
+            remindersForEachDeck.forEach { (did, reminders) ->
+                putString(DECK_SPECIFIC_KEY + did, Json.encodeToString(reminders))
+            }
+        }
+    }
+}
```
Can this be optimised? Avoid parsing the same SharedPreferences values twice and avoid the need to re-group later.

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,137 @@
+     */
+    private fun getRemindersForKey(key: String): MutableList<ReviewReminder> {
+        val jsonString = context.sharedPrefs().getString(key, null) ?: return mutableListOf()
+        return Json.decodeFromString<MutableList<ReviewReminder>>(jsonString)
```
What happens if the JSON string is corrupted? Use try-catch maybe

criticalAY reviewed (changes requested):
Solid niceee

ericli3690 commented:
Thanks Ashish! I've added some try-catch blocks to handle corrupt JSON strings and added a few unit tests to test that behaviour.

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,166 @@
+     * Edit all [ReviewReminder]s that are associated with a specific deck by operating on a single mutable list.
+     */
+    fun editAllDeckSpecificReminders(reminderEditor: (MutableList<ReviewReminder>) -> List<ReviewReminder>) {
+        val newReminders = reminderEditor(getAllDeckSpecificReminders())
```
`getAllDeckSpecificReminders()` internally decodes every JSON string in SharedPreferences, and then inside reminderEditor, the flat list is returned, and must be re-grouped again.
So:
- Decodes once, which is nice
-  But there's no grouped structure to begin with, and grouping happens only at the end, forcing repeated transformations.

What if we use this

```
 fun editAllDeckSpecificReminders(
        reminderEditor: (MutableList<ReviewReminder>) -> List<ReviewReminder>
    ) {
        val existingRemindersGrouped = getAllDeckRemindersGrouped()
        val updatedReminders = reminderEditor(existingRemindersGrouped.values.flatten().toMutableList())
        val updatedGrouped = updatedReminders.groupBy { it.did }

        val existingKeys = existingRemindersGrouped.keys.map { DECK_SPECIFIC_KEY + it }

        context.sharedPrefs().edit {
            existingKeys.forEach { remove(it) }
            updatedGrouped.forEach { (did, reminders) ->
                putString(DECK_SPECIFIC_KEY + did, Json.encodeToString(reminders))
            }
        }
    }


    private fun getAllDeckRemindersGrouped(): Map<DeckId, MutableList<ReviewReminder>> {
        return context.sharedPrefs().all
            .filterKeys { it.startsWith(DECK_SPECIFIC_KEY) }
            .mapNotNull { (key, value) ->
                val did = key.removePrefix(DECK_SPECIFIC_KEY).toLongOrNull() ?: return@mapNotNull null
                val reminders = try {
                    Json.decodeFromString<MutableList<ReviewReminder>>(value.toString())
                } catch (e: Exception) {
                    null
                } ?: return@mapNotNull null
                did to reminders
            }
            .toMap()
    }
```

Here, SharedPreferences read + decode only happens once, and the grouping logic is explicit and ig Less parsing, and a cleaner transformation chain. 

You may inspect more and reject this above suggestion if its not valid

ericli3690 replied:
Ah, I understand what you mean now. Sure, I'll do that, it is quite elegant.

ericli3690 commented:
Added getAllDeckSpecificRemindersGrouped. Refactored getAllDeckSpecificReminders and editAllDeckSpecificReminders to use it.

ericli3690 commented:
- Replaced `List` with `HashMap` to make editing individual review reminders more efficient
- Added two utility methods to `ReviewRemindersDatabase` to flatten and group maps of review reminders
- Updated `ReviewRemindersDatabaseTest`, which no longer needs to inefficiently sort lists of review reminders when asserting

ericli3690 commented:
- Added the `enabled` property to `ReviewReminder`

SanjaySargam commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,130 @@
+    val did: DeckId,
+    var enabled: Boolean,
+) {
+    init {
```
Use a single `require `with a compound condition

ericli3690 replied:
Hm, I think I'd actually like to keep the separate `require` lines if possible. Throwing everything together into a single compound condition makes the code a bit harder to read.

SanjaySargam replied:
I meant something like this : 
```kotlin
init {
    require(isValid()) { "Invalid ReviewReminder: hour=$hour, minute=$minute, snoozeTimeInterval=$snoozeTimeInterval=$snoozeTimeInterval=$snoozeAmountAllowed, cardTriggerThreshold=$cardTriggerThreshold" }
}

private fun isValid(): Boolean {
    return hour in (0..23) &&
           minute in (0..59) &&
           snoozeTimeInterval >= 0 &&
           cardTriggerThreshold >= 0
}
```

ericli3690 replied:
Wouldn't that make it so that if the `require` fails, it's hard to tell why it failed? I think if we break it up over multiple lines, and one particular condition fails, then we get a more clear and specific error message.

SanjaySargam commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,191 @@
+        val existingReminders = getRemindersForKey(key)
+        val updatedReminders = reminderEditor(existingReminders)
+        context.sharedPrefs().edit {
+            putString(key, Json.encodeToString(updatedReminders))
```
Should we add `try-catch` around Json.encodeToString() in case it fails, so SharedPreferences doesn’t get into a bad state?

ericli3690 replied:
Ah, good point. Sure, I'll add that in to be safe, though I won't be able to add unit tests for it. The type-checking of `reminderEditor`'s contract prevents me from triggering a serialization error from these editing methods. Thanks!

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,130 @@
+     * - [DISABLED]: The snooze button will never appear on notifications set by this review reminder.
+     * @see SnoozeAmount
+     */
+    object SpecialSnoozeAmounts {
```
Consider using a `sealed class` to make a union (serializability untested):

```kotlin
@Serializable
sealed class SnoozeAmount {
    data class Limited(val maxSnoozes: Int, val interval: Duration): SnoozeAmount()
    /** The snooze button will always be available */
    data object Infinite: SnoozeAmount()
    /** The snooze button will never appear */
    data object Disabled: SnoozeAmount()
}
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,130 @@
+@ConsistentCopyVisibility
+data class ReviewReminder private constructor(
+    val id: ReviewReminderId,
+    val hour: Int,
+    val minute: Int,
```
Consider combining `Hour` and `Minute` in a class, so invalid values can't be passed in

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,130 @@
+         */
+        private fun getAndIncrementNextFreeReminderId(context: Context): ReviewReminderId {
+            val nextFreeId =
+                context.sharedPrefs().getInt(
+                    NEXT_FREE_ID_KEY,
+                    FIRST_REMINDER_ID,
+                )
+            context.sharedPrefs().edit {
+                putInt(NEXT_FREE_ID_KEY, nextFreeId + 1)
+            }
```
As mentioned in a previous review, look into the `Prefs` class

david-allison commented on AnkiDroid/src/test/java/com/ichi2/anki/reviewreminders/ReviewReminderTest.kt:
```diff
@@ -0,0 +1,57 @@
+        activity = startRegularActivity<AnkiActivity>()
+    }
+
+    @After
+    override fun tearDown() {
+        super.tearDown()
+        // Reset the database after each test
+        activity.sharedPrefs().edit { clear() }
+    }
```
Is this necessary?

ericli3690 replied:
Technically, no, since there's only one test in this file right now. I added it in case someone else decides to write more tests for this file later, because they might forget to implement `tearDown()`.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,130 @@
+ * @param hour
+ * @param minute
+ * @param snoozeAmountAllowed Amount of times the user can hit snooze on the notification. See [SpecialSnoozeAmounts] for distinguished values.
+ * @param snoozeTimeInterval Time interval between snoozes, in minutes. Unused if snoozesAllowed is a [SpecialSnoozeAmounts] value.
```
Doh! Nope, you're right. Typo on my part.

ericli3690 commented:
Thanks for the reviews.

@SanjaySargam Regarding whether review reminders should remain in SharedPreferences after the corresponding deck is deleted: no, they should not, and don't worry, it's on my to-do list. Thanks for reminding me!

ericli3690 commented:
- Added try-catch for JSON encoding
- Created a `APP_WIDE_REMINDER_DECK_ID` constant
- Added `ReviewReminderTime` and `SnoozeAmount` classes
- Cleaned up references to `context.sharedPrefs()` to instead use `Prefs` or `AnkiDroidApp.sharedPrefs()`
- Added a new preferences resource to allow for `Prefs.getString()`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,172 @@
+         * @return The next free reminder ID.
+         */
+        private fun getAndIncrementNextFreeReminderId(): ReviewReminderId {
+            val nextFreeId = Prefs.getInt(R.string.review_reminders_next_free_id, FIRST_REMINDER_ID)
```
Look into `by intPref`. Use this new property as an opportunity to document the preference

```patch
Index: AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt b/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt	(revision 9d0e81e75ce77a0d7ce8949b5344af2318389b23)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt	(date 1750647615025)
@@ -16,7 +16,7 @@
 
 package com.ichi2.anki.reviewreminders
 
-import com.ichi2.anki.R
+import com.ichi2.anki.reviewreminders.ReviewReminder.Companion.APP_WIDE_REMINDER_DECK_ID
 import com.ichi2.anki.settings.Prefs
 import com.ichi2.libanki.DeckId
 import kotlinx.serialization.Serializable
@@ -131,11 +131,6 @@
          */
         const val APP_WIDE_REMINDER_DECK_ID = -1L
 
-        /**
-         * IDs start at this value and climb upwards by one each time.
-         */
-        private const val FIRST_REMINDER_ID = 0
-
         /**
          * Create a new review reminder. This will allocate a new ID for the reminder.
          * @return A new [ReviewReminder] object.
@@ -163,8 +158,8 @@
          * @return The next free reminder ID.
          */
         private fun getAndIncrementNextFreeReminderId(): ReviewReminderId {
-            val nextFreeId = Prefs.getInt(R.string.review_reminders_next_free_id, FIRST_REMINDER_ID)
-            Prefs.putInt(R.string.review_reminders_next_free_id, nextFreeId + 1)
+            val nextFreeId = Prefs.reviewReminderNextFreeId
+            Prefs.reviewReminderNextFreeId = nextFreeId + 1
             Timber.d("Generated next free review reminder ID: $nextFreeId")
             return nextFreeId
         }
Index: AnkiDroid/src/main/java/com/ichi2/anki/settings/Prefs.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/main/java/com/ichi2/anki/settings/Prefs.kt b/AnkiDroid/src/main/java/com/ichi2/anki/settings/Prefs.kt
--- a/AnkiDroid/src/main/java/com/ichi2/anki/settings/Prefs.kt	(revision 9d0e81e75ce77a0d7ce8949b5344af2318389b23)
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/settings/Prefs.kt	(date 1750647582518)
@@ -204,4 +204,6 @@
     var isDevOptionsEnabled: Boolean
         get() = getBoolean(R.string.dev_options_enabled_by_user_key, false) || BuildConfig.DEBUG
         set(value) = putBoolean(R.string.dev_options_enabled_by_user_key, value)
+
+    var reviewReminderNextFreeId by intPref(R.string.review_reminders_next_free_id, defaultValue = 0)
 }

```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,172 @@
+        /**
+         * Get and return the next free reminder ID which can be associated with a new review reminder.
+         * Also increment the next free reminder ID stored in SharedPreferences.
+         * Since there are 4 billion IDs available, this should not overflow in practice.
```
```suggestion
```

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,212 @@
+            // If, during development, the review reminders storage schema (i.e. ReviewReminder) is modified,
+            // this class will attempt to read review reminders stored in the old schema via the new schema, causing a SerializationException.
+            // To delete all review reminder keys in SharedPreferences, uncomment the following line.
+            // AnkiDroidApp.sharedPrefs().edit { AnkiDroidApp.sharedPrefs().all.keys.filter { it.startsWith("review_reminders_") }.forEach { remove(it) } }
```
Was this discussed and confirmed with mentors? We may want to use a separate file for these

* This will significantly increase the on-disk size of the default `SharedPreferences` -  maybe to the point of affecting app loading times
* We'll want to exclude this data from ACRA

criticalAY replied:
ACRA is where all the crash reports are sent, which is very helpful

ericli3690 replied:
@david-allison I've tried creating twenty decks and adding twenty review reminders to each. Launching the app, opening up the RecyclerView that displays all the review reminders etc. all works with no noticeable latency at all. If some user creates multiple hundred decks, each with their own large amounts of review reminders, _maybe_ there'll be a slowdown, but honestly I don't think most users will do that. Since I've already done a fair amount of work making review reminders work within SharedPreferences, I think I'll keep them there for now, unless you really think I should spend some time putting them into a separate file. If needed I can always go back in August and change this.

Please let me know if there's any action I should take regarding ACRA.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,212 @@
+    private fun decodeJson(jsonString: String): HashMap<ReviewReminderId, ReviewReminder> =
+        try {
+            // If, during development, the review reminders storage schema (i.e. ReviewReminder) is modified,
+            // this class will attempt to read review reminders stored in the old schema via the new schema, causing a SerializationException.
```
You'll want to plan how to update the schema

david-allison reviewed (commented):
All looks VERY well documented!

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,172 @@
+        /**
+         * The "deck ID" field for review reminders that are app-wide rather than deck-specific.
+         */
+        const val APP_WIDE_REMINDER_DECK_ID = -1L
```
Your timing could not have been more perfect. I was creating some dialog logic, suddenly realized I should have used a sealed class here, and then immediately got an email notifying me about a new review on my PR 😆

criticalAY commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,212 @@
+ * of cards due across all decks.
+ */
+class ReviewRemindersDatabase(
+    val context: Context,
```
Can we make this class context free as we are using the context only to show errors and that can be handled other way without context, see where it is going to be used and then utilise the error there

criticalAY reviewed (changes requested):
One more change

ericli3690 commented:
- Used an inline value class for `ReviewReminderId`, modified `ReviewReminderTest` accordingly
- Modified `ReviewReminderTime`'s methods to better use the Duration API
- Modified `ReviewReminderCardTriggerThreshold` to be an inline value class
- Moved `ReviewReminder` field types out of `ReviewReminder` so that they are no longer nested classes; this allows for less verbose code in other files
- `ReviewReminderScope` from `ScheduleReminders` is now used in `ReviewReminder` to indicate deck ID, it will be used in both places, as fundamentally indicating what scope is being edited by `ScheduleReminders` and which scope a `ReviewReminder` is attached to can be done by the same sealed data class; I personally think this is an elegant abstraction
- Used `by intPref` for the next free review reminder ID
- Shortened the docstring of `getAndIncrementNextFreeReminderId`
- Implemented the backend of the `ReviewReminder` migration process, including `attemptSchemaMigration` and `OldReviewReminderSchema`; the use of these backend methods from `ScheduleReminders` is also complete and has unit tests, but will be delivered in the upcoming RecyclerView PR
- Added an explanation of the `ReviewReminder` migration process to the docstring of `ReviewReminder`
- Added `getAllReviewReminderSharedPrefsAsMap` and `deleteAllReviewReminderSharedPrefs` helper methods to `ReviewRemindersDatabase` and added tests for them to `ReviewRemindersDatabaseTest`
- Cleaned up tests in `ReviewRemindersDatabaseTest` to minimize code duplication
- Made `ReviewRemindersDatabase` entirely Context-free; error-checking and dialog-throwing is complete in the form of `ScheduleReminders.catchDatabaseExceptions`, but will be delivered in the upcoming RecyclerView PR
- Added lines to `ReviewRemindersDatabase` docstrings indicating that methods may throw `SerializationExceptions`/`IllegalArgumentExceptions` and should be caught within the upcoming `ScheduleReminders.catchDatabaseExceptions` method

ericli3690 commented:
- Rebased on main
- Fixed typo: Added the missing `enabled` param to the docstring of `ReviewReminder`
- Tweaked a few tests to validate manually passing the optional `enabled` field to `createReviewReminder`

david-allison reviewed (commented):
Unsubmitted review, comments may not be current

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,226 @@
+        val maxSnoozes: Int,
+    ) : ReviewReminderSnoozeAmount() {
+        init {
+            require(timeInterval >= 0.minutes) { "Snooze time interval must be >= 0 minutes" }
```
Oops, fixed! We'll disallow zero.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,226 @@
+    ) : ReviewReminderSnoozeAmount() {
+        init {
+            require(timeInterval >= 0.minutes) { "Snooze time interval must be >= 0 minutes" }
+            require(maxSnoozes >= 0) { "Max snoozes must be >= 0" }
```
As above.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewReminder.kt:
```diff
@@ -0,0 +1,226 @@
+        @IgnoredOnParcel
+        private var cachedDeckName: String? = null
+
+        suspend fun getDeckName(): String = cachedDeckName ?: withCol { decks.name(did) }.also { cachedDeckName = it }
```
Better idea. The change doesn't affect anything I have here, so I've made a separate PR at https://github.com/ankidroid/Anki-Android/pull/18758. If you disagree with pulling the string resource out, feel free to close it!

ericli3690 commented:
- Moved getAndIncrementNewFreeReminderId to ReviewReminderId
- Modified `require` clauses of ReviewReminderSnoozeAmount
- Renamed `timeInterval` property of ReviewReminderSnoozeAmount to `interval`
- Added more documentation to the docstrings of ReviewReminderSnoozeAmount and `ReviewReminderScope.DeckSpecific.getDeckName`
- Rebased on main

criticalAY reviewed (approved):
LGTM! 
Don't want it to be blocked for it to be perfect as there will be changes in future.

Arthur-Milchior reviewed (approved):
Only nits, that can be done later

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,280 @@
+
+    /**
+     * Get the [ReviewReminder]s for a specific key.
+     * @throws SerializationException If the string is not a valid JSON string.
```
Ah, good point. I've updated the docstrings of the methods in ReviewRemindersDatabase and will push them at the next best opportunity.

ericli3690 commented on AnkiDroid/src/main/java/com/ichi2/anki/reviewreminders/ReviewRemindersDatabase.kt:
```diff
@@ -0,0 +1,280 @@
+     * @throws SerializationException If the string is not a valid JSON string.
+     * @throws IllegalArgumentException If the decoded object is not a HashMap<[ReviewReminderId], [ReviewReminder]>.
+     */
+    private fun getAllDeckSpecificRemindersGrouped(): Map<DeckId, HashMap<ReviewReminderId, ReviewReminder>> {
```
Yes, that's intentional. I only use a `HashMap` when a user is likely to call a `.get()` operation on it to guarantee O(1) fetch. There's currently nowhere in my code that requires getting an arbitrary element of a `Map<DeckId, HashMap<ReviewReminderId, ReviewReminder>>`. I only ever need to iterate through each element, so there's no need to make it a HashMap.

I do agree with you that `Map<DeckId, HashMap<ReviewReminderId, ReviewReminder>>` is extremely long and confusing to read. I considered making a typealias, but that felt like it would be even more confusing. Thanks for drawing my attention to `Deck` and `NoteType`, though, as making this a separate type class of its own makes a lot more sense. I'll hold off on making this optimization for now to ensure I stay on schedule, but I'll add this feature to my wishlist and tackle it in August if I get the chance. Thanks!
